module/input/PS3Controller/src/PS3Controller.cpp

```diff
@@ -0,0 +1,239 @@
+        L1             = 4,
+        R1             = 5,
+        SELECT         = 8,
+        START          = 9,
+        PS             = 10,
+        JOYSTICK_LEFT  = 11,
+        JOYSTICK_RIGHT = 12,
+        DPAD_UP        = 13,
+        DPAD_DOWN      = 14,
+        DPAD_LEFT      = 15,
+        DPAD_RIGHT     = 16
+    };
+    enum class Axis : uint8_t {
+        JOYSTICK_LEFT_HORIZONTAL = 0,
+        JOYSTICK_LEFT_VERTICAL,
+        L2,
+        JOYSTICK_RIGHT_HORIZONTAL,
+        JOYSTICK_RIGHT_VERTICAL,
+        R2,
+        ACCELEROMETER_X = 23,
+        ACCELEROMETER_Y,
+        ACCELEROMETER_Z
+    };
+
+#pragma pack(push, 1)
+    struct JoystickEvent {
+        uint32_t timestamp;  // event timestamp in milliseconds
+        int16_t value;       // value
+        uint8_t type;        // event type
+        uint8_t number;      // axis/button number
+    };
+    // Check that this struct is not cache alligned
+    static_assert(sizeof(JoystickEvent) == 8, "The compiler is adding padding to this struct, Bad compiler!");
+#pragma pack(pop)
+
+    using extension::Configuration;
+
+    using message::input::ps3controller::Accelerometer;
+    using message::input::ps3controller::CircleButton;
+    using message::input::ps3controller::CrossButton;
+    using message::input::ps3controller::DPadButton;
+    using message::input::ps3controller::L1Button;
+    using message::input::ps3controller::L2Trigger;
+    using message::input::ps3controller::LeftJoystick;
+    using message::input::ps3controller::LeftJoystickButton;
+    using message::input::ps3controller::PSButton;
+    using message::input::ps3controller::R1Button;
+    using message::input::ps3controller::R2Trigger;
+    using message::input::ps3controller::RightJoystick;
+    using message::input::ps3controller::RightJoystickButton;
+    using message::input::ps3controller::SelectButton;
+    using message::input::ps3controller::SquareButton;
+    using message::input::ps3controller::StartButton;
+    using message::input::ps3controller::TriangleButton;
+
+    PS3Controller::PS3Controller(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+
+        on<Configuration>("PS3Controller.yaml").then([this](const Configuration& config) {
+            // Only reconnect if we are changing the path to the device
+            if ((controller_path.compare(config["controller_path"]) != 0)
+                || (accelerometer_path.compare(config["accelerometer_path"]) != 0)) {
+                controller_path    = config["controller_path"];
+                accelerometer_path = config["accelerometer_path"];
+                connect();
+            }
+        });
+
+        on<Shutdown>().then([this] { disconnect(); });
+
+        // Details on Linux Joystick API
+        // https://www.kernel.org/doc/Documentation/input/joystick-api.txt
+        // Trigger when the joystick has an data to read
+        on<IO>(controller_fd, IO::READ).then([this] {
```

Comment:
Currently, if the joystick fails to connect in the `on<Configuration>` or reconnects in the on every this `on<IO>` won't pick it up. You should move this on statement into the `connect` function and .unbind() the previous reaction on reconnection. Otherwise it won't work.